README.md
# 📺 MAF SDK Command Line Interface

> Development automation for your MAF3 App projects

[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FJasperV%2Fmaf-cli.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2FJasperV%2Fmaf-cli?ref=badge_shield)

[![NPM](https://nodei.co/npm/maf-cli.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/maf-cli/)

#### 📖 Contents

- [Install](#-install)
- [Configuration](#️-configuration)
- [Usage](#️-usage)
- [MAF key combinations](#️-maf-key-combinations)
- [About](#about)
- [MAF3-SDK](#-maf3-sdk)


## 💾 Install

```sh
$ npm i -g maf-cli
$ npm i -D maf-cli Metrological/maf3-sdk
$ touch .mafrc.js
$ maf init
$ maf --tasks
```

After running `maf init` all requirements will be created, installed and scaffolded based on the answers you provide. Some tasks have a dependency on a Metrological Dashboard API key. You can get one [here](https://dashboard.metrological.com/#/profile/api).

To use this module with an already existing app, simply move that app out of the MAF3 SDK folder into it's own and run `maf init` on it.

As a shorthand you could run `curl -L  git.io/maf-cli | bash` from within your project folder. And then run `maf init`.

### Moving from maf3-sdk to maf-cli

If you are accustomed to working with the maf3-sdk then working with the maf-cli should be transparent. A lot of things are now handled for you. You can now depend your project on the maf-cli vs. running your app within the maf3-sdk. 
All apps from the maf3-sdk are included. As well as any dependencies.
Your index.html file is no longer required. Any configuration you might have in there can be copied over to your maffile (.mafrc.js). Declaring your app, categories and ui is no longer required.


## ⚙️ Configuration

You can configure the MAF CLI via your .mafrc.js file. It has the following (default) options available:

```javascript
{
  language: `nl` // set the default language of the Store and App
, es6: true // when true all your code will be transpiled (via the Closure Compiler) to ES5 before running and publishing
, autostart: false // when true auto-starts your app when browsing to https://localhost:8443
, store: `YOUR/STORE/ID_HERE` // use this when running the store task for example: metrological/nl/external
}
```

Additionally, you are able to define your own tasks within your maffile.

```javascript
const maf = require( 'maf-cli' )

module.exports = {
  language: `nl`
, function hello() {
    console.log( `world` )
  }
}
```

This can be used to integrate any other tools you might use, or preprocess certain files.

You may also set some environment variables:

```sh
LANG=en_US.utf8 # for compatibility reasons, remove if this causes issues for you
TERM=xterm-256color # for compatibility reasons, remove if this causes issues for you
NODE_ENV=production # let's make sure we are equal to production as much as possible
NODE_PORT=8080 # change to your liking
NODE_SECURE_PORT=8443 # change to your liking
DOCS_PORT=9090 # change to your liking
METROLOGICAL_API_KEY="YOUR_API_KEY_HERE" # this is used for publishing your App
```


## ▶️ Usage

```sh
$ maf
```

The default task (run) will run and starts the [SDK Server](https://localhost:8443).

To run individual tasks, use `maf <task>`.

To display the list of build-in tasks, use `maf --tasks`

For an overview of available options and commands, use `maf --help`


## ⌨️ MAF key combinations:

| Command | Key |
| --- | --- |
| `stop` | Shift + Down |
| `play/pause` | Shift + Up |
| `rewind` | Shift + Left |
| `forward` | Shift + Right |
| `back` | Backspace |
| `red` | F1 |
| `green` | F2 |
| `yellow` | F3 |
| `blue` | F4 |
| `channel up` | PageUp |
| `channel down` | PageDown |


## ❌ Issue Submission

Please use GitHub Issues for issue submission.

### Known Issues

The build-in proxy service uses xml2json which uses node-expat. On Windows this requires some extra work setting up. Please refer to it's [documentation](https://www.npmjs.com/package/xml2json).


## ✅ Feature requests

Feature requests can be submitted via Issues as well.


## Contribute

Feel free to send in any pull requests.


## About

The idea for this module came from the desire to take the development of MAF3 apps into the 21st century. To allow for anyone to focus on building the app and less on managing releases and configuring the runtime environment. This module allows for your project to easily depend on the MAF3-SDK and make CI and CD more simple to handle. This is the first iteration of many more to come.


## 🛠 MAF3-SDK

You can learn more about the MAF3-SDK at it's [website](https://mafsdk.tv/), or on [GitHub](https://git.io/maf3-sdk).

For any questions regarding App developement please use the [Metrological Helpdesk](https://metrological.atlassian.net/servicedesk/customer/portal/).


## ©️ License

MIT


[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FJasperV%2Fmaf-cli.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2FJasperV%2Fmaf-cli?ref=badge_large)


CHANGELOG.md
<a name="1.7.3"></a>
## [1.7.3](https://github.com/JasperV/maf-cli/compare/1.7.2...1.7.3) (2018-07-16)


### Bug Fixes

* **compiler:** remove logs ([ca97876](https://github.com/JasperV/maf-cli/commit/ca97876))



<a name="1.7.2"></a>
## [1.7.2](https://github.com/JasperV/maf-cli/compare/1.7.1...1.7.2) (2018-07-16)


### Bug Fixes

* **compiler:** inclusion bug ([829aaf5](https://github.com/JasperV/maf-cli/commit/829aaf5))



<a name="1.7.1"></a>
## [1.7.1](https://github.com/JasperV/maf-cli/compare/1.7.0...1.7.1) (2018-07-09)


### Bug Fixes

* **release:** fix release api call ([978834c](https://github.com/JasperV/maf-cli/commit/978834c))


### Features

* add notifier ([1340f74](https://github.com/JasperV/maf-cli/commit/1340f74))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/JasperV/maf-cli/compare/1.6.0...1.7.0) (2018-06-04)


### Bug Fixes

* **compiler:** fix default es6 code input ([5b893fc](https://github.com/JasperV/maf-cli/commit/5b893fc))
* **init:** fix identifier in metadata ([581ae4c](https://github.com/JasperV/maf-cli/commit/581ae4c))
* **init:** fix init params and env params ([8158bf4](https://github.com/JasperV/maf-cli/commit/8158bf4))
* fix localization casing ([7508a0b](https://github.com/JasperV/maf-cli/commit/7508a0b))
* **package:** add es6/compiler to package task ([340057c](https://github.com/JasperV/maf-cli/commit/340057c))
* **server:** do not clear screen when starting ([6e27431](https://github.com/JasperV/maf-cli/commit/6e27431))
* **server:** fix wrong source order ([7a38ab9](https://github.com/JasperV/maf-cli/commit/7a38ab9))
* **server:** remove minify html directive ([d8fe721](https://github.com/JasperV/maf-cli/commit/d8fe721))
* **sync:** fix identifier syncing ([90678f7](https://github.com/JasperV/maf-cli/commit/90678f7))


### Features

* **compiler:** add prettier ([3fbd2ee](https://github.com/JasperV/maf-cli/commit/3fbd2ee))
* **package:** add spinner to package ([860f8a8](https://github.com/JasperV/maf-cli/commit/860f8a8))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/JasperV/maf-cli/compare/1.5.10...1.6.0) (2018-05-30)


### Bug Fixes

* **cli:** allow --version to run without local module ([5ab425a](https://github.com/JasperV/maf-cli/commit/5ab425a))



<a name="1.5.10"></a>
## [1.5.10](https://github.com/JasperV/maf-cli/compare/1.5.9...1.5.10) (2018-05-30)


### Bug Fixes

* fix require of path ([63748ed](https://github.com/JasperV/maf-cli/commit/63748ed))



<a name="1.5.8"></a>
## [1.5.8](https://github.com/JasperV/maf-cli/compare/1.5.7...1.5.8) (2018-05-30)


### Bug Fixes

* **server:** fix path off certificate and key ([8fed619](https://github.com/JasperV/maf-cli/commit/8fed619))



<a name="1.5.7"></a>
## [1.5.7](https://github.com/JasperV/maf-cli/compare/1.5.6...1.5.7) (2018-05-28)


### Bug Fixes

* maf3-sdk dependency warning ([ffa6113](https://github.com/JasperV/maf-cli/commit/ffa6113))
* **installation:** fix path for translations file ([f3bd6bd](https://github.com/JasperV/maf-cli/commit/f3bd6bd))



<a name="1.5.6"></a>
## [1.5.6](https://github.com/JasperV/maf-cli/compare/1.5.5...1.5.6) (2018-05-25)


### Bug Fixes

* **installation:** fix path in certificate check ([2793c26](https://github.com/JasperV/maf-cli/commit/2793c26))



<a name="1.5.5"></a>
## [1.5.5](https://github.com/JasperV/maf-cli/compare/1.5.4...1.5.5) (2018-05-25)


### Bug Fixes

* **npm:** fix installation via npm ([2a2a2ff](https://github.com/JasperV/maf-cli/commit/2a2a2ff))



<a name="1.5.4"></a>
## [1.5.4](https://github.com/JasperV/maf-cli/compare/v1.5.3...v1.5.4) (2018-05-25)



<a name="1.5.0"></a>
# [1.5.0](https://github.com/JasperV/maf-cli/compare/1.4.0...1.5.0) (2018-05-17)


### Features

* test new way of install ([83308a7](https://github.com/JasperV/maf-cli/commit/83308a7))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/JasperV/maf-cli/compare/1.3.18...1.4.0) (2018-04-25)





lib/server.js
#!/usr/bin/env node

'use strict'

const os = require( `os` )
const fs = require( `fs` )
const path = require( 'path' )
const http = require( 'http' )
const https = require( 'https' )
const ngrok = require( 'ngrok' )
const clear = require( 'clear' )
const morgan = require( 'morgan' )
const qs = require( 'querystring' )
const express = require( 'express' )
const compression = require( 'compression' )
const requestStats = require( 'request-stats' )

const ansi = require( './util/ansi' )
const hosts = require( './util/hosts' )( { family: `IPv4` } )

const port = process.env.NODE_PORT || 8080
const securePort = process.env.NODE_SECURE_PORT || 8443

const app = express()
const counterApp = express()

let firstRequest = false

http.globalAgent.maxSockets = Infinity
https.globalAgent.maxSockets = Infinity

function getRemoteHost( port ) {
  return ngrok.connect( {
    addr: port
  , region: `eu`
  } )
}

function startServer( config ) {
  let url
  const secure = config.secure

  app.disable( `x-powered-by` )

  app.use( morgan( `dev` ) )
  app.use( compression() )

  app.use( require( './routes/index' )( config.store, config.meta ) )
  app.use( require( './routes/ip' ) )
  app.use( require( './routes/proxy' ) )

  app.use( ( req, res ) => res.status( 404 ).send( `404 - Not Found` ) )

  app.use( ( err, req, res ) => {
    console.error( err.stack )
    res.status( 500 ).send( `500 - Internal Server Error` )
  } )

  const serverPort = secure ? securePort : port
  const counterPort = secure ? port : securePort
  const protocol = secure ? `https` : `http`
  const counterProtocol = secure ? `http` : `https`
  const httpsOptions = {
    key: fs.readFileSync( path.resolve( process.cwd(), `MAF3SDK.key` ), `utf8` )
  , cert: fs.readFileSync( path.resolve( process.cwd(), `MAF3SDK.crt` ), `utf8` )
  }

  counterApp.get( `*`, ( req, res ) => {
    res.redirect(
      302
    , `${req.protcol === protocol ? counterProtocol : protocol}://${req.headers.host.split( `:` ).shift()}:${serverPort}${req.path}?${qs.stringify( req.query )}`
    )
  } )

  const server = secure ? https.createServer( httpsOptions, app ) : http.createServer( app )
  const counterServer = secure ? http.createServer( counterApp ) : https.createServer( httpsOptions, counterApp )

  require( './util/terminus' )( server )

  requestStats( server, stats => {
    if ( firstRequest ) return
    firstRequest = true
    clear()
  } )

  server.listen( serverPort, async _ =>
    process.send( {
      serverStart: {
        secure
      , hosts
      , port: serverPort
      , external: config.external ? await getRemoteHost( serverPort ) : false
      }
    } )
  )

  counterServer.listen( counterPort )
}

process.on( `message`, m => {
  if ( m.start ) startServer( m )
} )



/*

  Was used on sending json via post to api server

  bodyParser = require('body-parser'),

  var urlParser = bodyParser.urlencoded({extended: false, limit: '5mb'});
  var rawParser = function(req, res, next) {
    req.rawBody = '';
    req.on('data', function(chunk) {
      req.rawBody += chunk;
    });

    req.on('end', function() {
      try{
        if(req.headers['content-type'] === 'application/json')
          req.body = JSON.parse(req.rawBody);
      }catch(e){}
      return next();
    });
  };
*/
